pkg/controller/workspace/tfc_run.go

```diff
@@ -29,6 +29,24 @@ func (t *TerraformCloudClient) UploadConfigurationFile(uploadURL string) error {
 	return nil
 }
 
+// WaitForConfigurationUploaded waits a small amount of time for the configuration version to be "uploaded"
+func (t *TerraformCloudClient) WaitForConfigurationUploaded(id string) error {
```

Comment:
I wanted to solve a similar problem for VCS backed workspaces and I was told at the time it's better to avoid blocking in the reconcile function and just retry instead. See https://github.com/hashicorp/terraform-k8s/pull/70#discussion_r498518775 for more context.

As a result I changed my method to return a bool if we need to requeue (and retry) of if we were able to start the job. See https://github.com/hashicorp/terraform-k8s/blob/80bf499e93541db07e780b55ecdb955e6ea5b797/pkg/controller/workspace/workspace_controller.go#L403-L407 and 
https://github.com/hashicorp/terraform-k8s/blob/80bf499e93541db07e780b55ecdb955e6ea5b797/pkg/controller/workspace/workspace_controller.go#L374-L390 .


That said and while I'm typing these words I realise that the only thing `prepareVcsRun` does is to check if there is a state available or not. Since we need this check for both cases (VCS or module based) a possible solution would be to rename `prepareVcsRun` to `checkForConfigVersion` and change the if block from this:
```
	var err error
	requeue := false
	if instance.Spec.VCS != nil {
		requeue, err = r.prepareVCSRun(instance)
	} else if instance.Spec.Module != nil {
		err = r.prepareModuleRun(instance, options)
	}
	if err != nil {
		return err
	} else if requeue {
		// When requeue is true it means that the workspace is not ready yet.
		return nil
	}
```

to this: 
```
	if instance.Spec.Module != nil {
		if err := r.prepareModuleRun(instance, options); err != nil {
                        return err
                }
	}

        requeue, err = r.checkForConfigVersion(instance)
	if err != nil {
		return err
	} else if requeue {
		// When requeue is true it means that the workspace is not ready yet.
                // Since we are returning before setting a RunID the reconcile method will 
                // retry this part of the code.
		return nil
	}

```

(I wrote the code in the PR editor, there are probably a few typos here and there :) ) .

You may want to enhance `checkForConfig` (currently named `prepareVcsRun`) to optionally look for a specific state ID instead of just checking if there is a state uploaded or not.